core/src/message/context.rs

```diff
@@ -278,20 +278,23 @@ impl MessageContext {
         packet: InitPacket,
         delay: u32,
     ) -> Result<(MessageId, ProgramId), Error> {
-        let program_id = packet.destination();
-
-        if self.store.initialized.contains(&program_id) {
-            return Err(Error::DuplicateInit);
-        }
-
         let last = self.store.outgoing.len() as u32;
 
         if last >= self.settings.outgoing_limit {
             return Err(Error::OutgoingMessagesAmountLimitExceeded);
         }
 
         let message_id = MessageId::generate_outgoing(self.current.id(), last);
-        let message = InitMessage::from_packet(message_id, packet);
+
+        // Not sure if this is correct, it seems much more likely to generate duplicates
+        // if the same program is uploaded twice from the same parent program.
+        let message = InitMessage::from_packet(self.current.id(), packet);
```

Comment:
wrong `message_id` is used to create `InitMessage`
```suggestion
        let message = InitMessage::from_packet(message_id, packet);
```